File: palindrome.py
# Program to check if a number is a palindrome

#=====================================================================================
#                  Method1: While loop
#=====================================================================================

def isPalindrome1( number ):
    num     = number                          # Assign value of number to num
    reverse = 0                               # Assign 0 to reverse
    
    # Negative integers are mot palindromic
    if ( number < 0 ):
        return False
    
    # loop until num has no digit but 0
    while ( num != 0 ):
        reverse = (10 * reverse) + (num % 10) # adding last digit of num to reverse
        num   //= 10                          # remove last digit from num

    return (number == reverse)

#=====================================================================================
#                Method2: For loop with arrays
#=====================================================================================

def isPalindrome2( number ):
    reverse = 0                               # Assign 0 to reverse

    # Negative integers are mot palindromic
    if ( number < 0 ):
        return False

    div_by_base10 = [ number // (1*(10**i)) for i in range(len(str(number))) ]
    last_digits   = [ i % 10 for i in div_by_base10 ]

    for last_digit in last_digits:
        reverse = 10 * reverse + last_digit

    return (number == reverse)

#=======================================================================================
#              Method3: recursion, reverse of an integer
#=======================================================================================

def reverse( number ):
    return reverseNumber(number, 0)

def reverseNumber(number, reverse):
    if (number == 0):
        return reverse
    return reverseNumber(number // 10, 10 * reverse + number % 10)

def isPalindrome3( number ):
    return (number == reverse(number) if number > 0 else False)

#=======================================================================================
#            Method4: string reverse
#=======================================================================================

def isPalindrome4( number ):
    number = str(number)
    return ( number == number[::-1] if int(number) > 0 else False)

#=======================================================================================
#           Main
#=======================================================================================

if __name__ == "__main__":

    number = int(input("Enter a number: "))
    print()

    print("While Loop: {} is a palindrome.".format(number) if isPalindrome1(number) 
            else "While loop: {} is not a palindrome.".format(number))
    print()
    print("For Loop (w/ arrays): {} is a palindrome.".format(number) if isPalindrome2(number)
            else "For loop (w/ arrays): {} is not a palindrome.".format(number))
    print()
    print("Recursion (Reverse of a number): {} is a palindrome.".format(number) if isPalindrome3(number)
            else "Recursion (Reverse of a number): {} is not a palindrome.".format(number))
    print()
    print("String reverse: {} is a palindrome.".format(number) if isPalindrome4(number)
            else "String reverse: {} is not a palindrome.".format(number))

File: palindrome/src/palindrome.cpp
/*
 * Program to check if a given number is a palindrome.
 */

#include <iostream>

using namespace std;

//===================================================================
//			    FIRST METHOD
//===================================================================

bool isPalindrome1(int number)
{
	int num     = number;
	int reverse = 0;
	
	// negative numbers are not palindromes
	if ( number < 0 )
		return false;

	while ( num != 0 )
	{
		// Add the last digit of num to reverse
		reverse = (10 * reverse) + (num % 10);
		// remove the last digit of num
		num    /= 10;
	}

	return ( number == reverse );
}

//===================================================================
//                         SECOND METHOD
//===================================================================
int reverseNumber1(int number, int reverse);
int reverse1( int number )
{
	return reverseNumber1(number, 0);
}
int reverseNumber1(int number, int reverse)
{
	// Reverse of a one digit number is that number
	if ( number == 0 ) return reverse;
	// Add last digit of number to reverse and remove from number
	return reverseNumber1( number / 10, 10 * reverse + number % 10);
}
bool isPalindrome2(int number)
{
	return ( number == reverse1(number));
}
//===================================================================
//			  THIRD METHOD
//===================================================================

string reverseNumber2( string number )
{
	if ( number.size() == 1 )
		return number;
	return number[number.size()-1] + reverseNumber2( number.substr(0, number.size()-1) );
}
bool isPalindrome3(int number)
{
	return (number == stoi(reverseNumber2(to_string(number))));
}
int main()
{
	int number;
	cout << "Enter a number: ";
	cin  >> number;
	
	
	if (isPalindrome1(number))
		cout << "Method1: " << number << " is a palindrome." << endl;
	else
		cout << "Method1: " << number << " is not a palindrome." << endl;
	
	  if (isPalindrome2(number))
                cout << "Method2: " << number << " is a palindrome." << endl;
        else
                cout << "Method2: " << number << " is not a palindrome." << endl;

	  if (isPalindrome3(number))
                cout << "Method3: " << number << " is a palindrome." << endl;
        else
                cout << "Method3: " << number << " is not a palindrome." << endl;

}
